### ISSUE TITLE
Allow configuration with lazy project-relative path

### ISSUE THREAD

**Issue by dansanduleac** [jshiell/checkstyle-idea#333]
### Problem

Currently, if I try to add a checkstyle config in the settings, it doesn't let me actually type a relative path, whether or not I tick the "Store relative to project location".

### Example

Here, I'm trying to set up a default checkstyle configuration under `File -> Other Settings -> Default Settings...` (as opposed to per-project settings). Same problem occurs under project settings.

![image](https://user-images.githubusercontent.com/222827/28379531-e4fbb018-6cab-11e7-854b-753d4297750a.png)
Result:

![image](https://user-images.githubusercontent.com/222827/28379538-ef50425e-6cab-11e7-8a5a-a2f4950e1221.png)

This is very frustrating as every time I have a lot of projects with checkstyle configured in the same location, but I have to configure each one individually, and also navigate to / find the full path of the checkstyle file.

### Proposed solution

I'd like to be able to put a lazy path in there, that is later interpreted as relative to the current IDEA project.

**Comment by jshiell** [comment]
`Store relative to project location` only relates to how we store the value. You'd still need to browse to the file originally.

However, your suggested solution does seem sensible - to treat the base dir as relative to the project, not the runtime CWD. I'll add it to the todo list; no promises on timeframe, however.

**Comment by jshiell** [comment]
We'll now try and resolve any entered file name relative to the project directory (assuming it can't be found as an absolute path).

Note that this isn't lazy, as we immediately check the rules file for validity (and read the properties). If you need a lazy reference, your best source is the Eclipse-CS properties, e.g. `project_loc`.

**Comment by dansanduleac** [comment]
I'd like to revisit the lazy setup.
Even with the change, without this being lazy, I can't set a default checkstyle path, because it never goes through the code that resolves it against the project path.

It should just be resolved every time the rules are _loaded_ as opposed to when they are _created_.

**Comment by dansanduleac** [comment]
Here's what happens when I create a new project, having the default relative path I mentioned in the description:

![rules-loaded-from-default](https://user-images.githubusercontent.com/222827/28518023-670eab68-705e-11e7-8354-22081688e714.gif)

I get this constant Loading in the middle, I can't click edit, and have a bunch of these errors in the event log:
```
10:47 AM	Properties could not be read from the Checkstyle configuration file
```
which repeat periodically.

**Comment by jshiell** [comment]
I fear the lazy evaluation of the filename isn't going to happen at present, as it would break the caching structure, which assumes a configuration location is static (i.e. 'Baseline checkstyle' will be equally valid every time it's loaded). To change this would require reworking the caching to ensure we don't pollute the cache. 

From the description above, though, I assume you're actually interested in these being evaluated when being copied from the default project, rather than every time? In which case it may be simpler. I'll add it to the list; as always, no promises on when/if.

**Comment by dansanduleac** [comment]
For my use case the latter will suffice! :)

Even if it were the former, instead of having `path/to/checkstyle.xml` be part of the caching key, we could have `${PROJECT_DIR}` and the project-relative path be part of the key instead. That way if the project moves location, the cache is invalidated, otherwise not.

**Comment by jshiell** [comment]
For the former, the problem is not so much changing the cache key (or even ensuring the caches are project specific, which is probably the easiest solution). The bigger problem is that a lot of the plugin code was written a long time ago, when I was young and stupid (or at least stupider). Hence there's not a lot of tests, and even small changes can be risky. It all seems a good idea at the time, and haunts me now 😄

**Comment by dansanduleac** [comment]
Ping @jshiell :) Still having to configure this manually on every new project I open.
Any idea if you'll have time to work on this?

**Comment by jshiell** [comment]
I fear it won't be until the end of September at the earliest, as my in-laws are over from NZ and we're spending a lot of time bouncing around with them. But feel free to keep pinging me 😄

**Comment by dansanduleac** [comment]
Ping again @jshiell :)

**Comment by jshiell** [comment]
Thanks for the reminder :-) I'm gainfully unemployed for a few days later in November, so hoping to be able to spend a bit of time catching up on my backlog!

**Comment by dansanduleac** [comment]
Alright! :) Thanks for looking into it.

**Comment by jshiell** [comment]
After a long delay, here's a first cut. It cut deeper than I expected, but I think I've got something working.

* In default project settings I enter `checkstyle.xml` (or `./checkstyle.xml`) and tick `store relative to project location`.
* I create a new project and stick `checkstyle.xml` in the root.
* In settings we now have the `checkstyle.xml` file with the correct path.
* Scanning with it works as expected 😄 

I'd really appreciate it if you could give this a quick bash and let me know if a) it works for you and b) if it does what you were after. Please note this build hasn't had a huge amount of testing. Thank you!

https://www.dropbox.com/s/l0bksb42zjnfovg/CheckStyle-IDEA-5.12.2.zip?dl=0

**Comment by jshiell** [comment]
Released in 5.13.0.
